Sources/SwiftGRPCNIO/GRPCClientChannelHandler.swift

```diff
@@ -0,0 +1,200 @@
+    case expectingHead
+    case expectingMessageOrEnd
+    case ignore
+  }
+
+  private var inboundState: InboundState = .expectingHeaders
+  private var outboundState: OutboundState = .expectingHead
+
+  /// Creates a new `GRPCClientChannelHandler`.
+  ///
+  /// - Parameters:
+  ///   - initialMetadataPromise: a promise to succeed on receiving the initial metadata from the service.
+  ///   - statusPromise: a promise to succeed with the outcome of the call.
+  ///   - responseObserver: an observer for response messages from the server; for unary responses this should
+  ///     be the `succeedPromise` case.
+  internal init(
+    initialMetadataPromise: EventLoopPromise<HTTPHeaders>,
+    statusPromise: EventLoopPromise<GRPCStatus>,
+    responseObserver: ResponseObserver<ResponseMessage>
+  ) {
+    self.initialMetadataPromise = initialMetadataPromise
+    self.statusPromise = statusPromise
+    self.responseObserver = responseObserver
+  }
+
+  /// Observe the given status.
+  ///
+  /// The `status` promise is __succeeded__ with the given status despite `GRPCStatus` being an
+  /// `Error`. If `status.code != .ok` then the initial metadata and response promises (if applicable)
+  /// are failed with the given status.
+  ///
+  /// - Parameter status: the status to observe.
+  internal func observeStatus(_ status: GRPCStatus) {
+    if status.code != .ok {
+      self.initialMetadataPromise.fail(error: status)
+      self.responsePromise?.fail(error: status)
+    }
+    self.statusPromise.succeed(result: status)
+  }
+}
+
+extension GRPCClientChannelHandler: ChannelInboundHandler {
+  public typealias InboundIn = GRPCClientResponsePart<ResponseMessage>
+
+  /// Reads inbound data.
+  ///
+  /// On receipt of:
+  /// - headers: the initial metadata promise is succeeded.
+  /// - message: the message observer is called with the message; for unary responses a response
+  ///   promise is succeeded, otherwise a callback is called.
+  /// - status: the status promise is succeeded; if the status is not `ok` then the initial metadata
+  ///   and response promise (if available) are failed with the status. The channel is then closed.
+  public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+    guard self.inboundState != .ignore else { return }
+
+    switch unwrapInboundIn(data) {
+    case .headers(let headers):
+      guard self.inboundState == .expectingHeaders else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      self.initialMetadataPromise.succeed(result: headers)
+      self.inboundState = .expectingMessageOrStatus
+
+    case .message(let message):
+      guard self.inboundState == .expectingMessageOrStatus else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      self.responseObserver.observe(message)
+
+    case .status(let status):
+      guard self.inboundState == .expectingMessageOrStatus else {
```

Comment:
Will this work in case of a "Trailers-Only" response? (Not sure if the server only sends headers or trailers in that case.)

Reply:
Good question. For "Trailers-Only" the status will be sent in a single frame, so I assume this would fail.

Reply:
I'd suggest having a `TODO` (or even `FIXME`) to support "Trailers-Only" responses, then.

Reply:
It's supported in this revision.

Reply:
Sorry, I should have taken a closer look.